### ISSUE TITLE
[CSA] False negative: null argument report missed at third unknown-loop iteration

### ISSUE THREAD

**Issue by lynae99** [llvm/llvm-project#214232]
Hi, I found a false negative in `core.NonNullParamChecker` / `unix.cstring.NullArg` loop path exploration.

## Program

```c
#include <string.h>

volatile size_t observed_length;

void check(int count) {
  int i = 0;
  while (i < count) {
    if (i == 2)
      observed_length = strlen((char *)0);
    ++i;
  }
}
```

## Actual result

CSA emits no `core.NonNullParamChecker` or `unix.cstring.NullArg` warning:

```sh
clang --analyze \
  -Xclang -analyzer-checker=core.NonNullParamChecker,unix.cstring.NullArg \
  unknown_loop_third_iteration_nullarg.c
```

Changing only the loop condition from `i < count` to `i < 3` produces the expected warning. With the unknown bound, equivalent checks at `i == 0` and `i == 1` are reported, while `i == 2` and later are missed.

## Comparison with related programs

The missed program contains a null argument guarded by an unknown-bound loop:

```c
while (i < count) {
  if (i == 2)
    observed_length = strlen((char *)0); // no warning
  ++i;
}
```

Equivalent earlier-iteration checks are reported:

```c
while (i < count) {
  if (i == 0)
    observed_length = strlen((char *)0); // warning
  ++i;
}
```

Keeping the `i == 2` condition but changing the bound to the constant `3` also reports the diagnostic:

```c
while (i < 3) {
  if (i == 2)
    observed_length = strlen((char *)0); // warning
  ++i;
}
```

## Runtime confirmation

Building the provided file with `-DRUNTIME -fno-builtin-strlen -fsanitize=address` and executing `check(3)` produces an AddressSanitizer null-read crash in `strlen`.

This does not appear to be the inlined defensive-check suppression heuristic: `-analyzer-config suppress-null-return-paths=false` does not recover the diagnostic. It also does not appear to be a simple node-budget issue: `-analyzer-config max-nodes=0` does not recover it.

## Expected result

The path `count >= 3` is feasible and `i == 2` makes the null call definite. It would be useful for CSA to report the null argument independently of whether the loop bound is a parameter or the constant 3, or to make the loop-exploration tradeoff explicit.

## Verification

```text
clang version 24.0.0git (https://github.com/llvm/llvm-project.git 1cb7e838cd47ecad4050948c0c907ecb1f466ac3)
```

**Comment by NagyDonat** [comment]
This is intentional behavior of the analyzer which was implemented by my commit https://github.com/llvm/llvm-project/commit/bb27d5e5c6b194a1440b8ac4e5ace68d0ee2a849 .

The analyzer only emits bug reports when it has _justification_ to claim that the bug can actually occur. If the loop condition is opaque (the analyzer cannot deduce whether it is true or false), then the analyzer will assume two iterations (because if the programmer only intended at most one iteration, then they wouldn't have written a loop), but it doesn't assume a third iteration, because a loop intended for at most two iterations looks like a generic loop.

This may sound like theoretical hair-splitting, but assuming the third iteration in opaque loops can produce _lots_ of false positives: for example on the FFMPEG project (a stable open-source codebase) it produced 100+ false positive `security.ArrayBound` reports. This amount of noise (on reasonable code) is unacceptable for projects that want to use the analyzer in production.

If somebody wants to introduce an analysis mode that assumes third (and perhaps later) iterations in loops with opaque conditions, then I'm not opposed to merging it as an off-by default flag, but I don't think that it will be useful in practice and won't accept it as default.
